src/main/scala/inox/ast/ExprOps.scala

```diff
@@ -65,30 +65,80 @@ trait ExprOps extends GenTreeOps {
-        val res = super.transform(vd).freshen.toVariable
-        subst(vd.toVariable) = res
-        res
-      }).toVal
+    override def transform(id: Identifier, env: Env): Identifier = {
+      env.getOrElse(id, id)
+    }
 
-      override def transform(expr: Expr): Expr = expr match {
-        case v: Variable => transform(v.toVal).toVariable
-        case Choose(res, pred) if !freshenChooses =>
-          val newVd = super.transform(res)
-          subst(res.toVariable) = newVd.toVariable
-          Choose(newVd, transform(pred)).copiedFrom(expr)
-        case _ => super.transform(expr)
-      }
-    }.transform(expr)
+    override def transform(e: Expr, env: Env): Expr = e match {
+      case Let(vd, v, b) =>
+        val freshVd = vd.freshen
+        Let(transform(freshVd, env), transform(v, env), transform(b, env.updated(vd.id, freshVd.id))).copiedFrom(e)
+
+      case Lambda(params, b) =>
+        val (sparams, senv) = params.foldLeft((Seq[t.ValDef](), env)) {
+          case ((sparams, env), vd) =>
+            val freshVd = vd.freshen
+            (sparams :+ transform(freshVd, env), env.updated(vd.id, freshVd.id))
+        }
+        Lambda(sparams, transform(b, senv)).copiedFrom(e)
+
+      case Forall(params, b) =>
+        val (sparams, senv) = params.foldLeft((Seq[t.ValDef](), env)) {
+          case ((sparams, env), vd) =>
+            val freshVd = vd.freshen
+            (sparams :+ transform(freshVd, env), env.updated(vd.id, freshVd.id))
+        }
+        Forall(sparams, transform(b, senv)).copiedFrom(e)
+
+      case Choose(vd, pred) if !freshenChooses =>
+        Choose(transform(vd, env), transform(pred, env)).copiedFrom(e)
+
+      case Choose(vd, pred) if freshenChooses =>
+        val freshVd = vd.freshen
+        Choose(transform(freshVd, env), transform(pred, env.updated(vd.id, freshVd.id))).copiedFrom(e)
+
+      case _ =>
+        super.transform(e, env)
+    }
+
+    override def transform(tpe: s.Type, env: Env): t.Type = tpe match {
+      case RefinementType(vd, prop) =>
+        val freshVd = vd.freshen
+        RefinementType(transform(freshVd, env), transform(prop, env.updated(vd.id, freshVd.id))).copiedFrom(tpe)
+
+      case PiType(params, to) =>
+        val (sparams, senv) = params.foldLeft((Seq[t.ValDef](), env)) {
+          case ((sparams, env), vd) =>
+            val freshVd = vd.freshen
+            (sparams :+ transform(freshVd, env), env.updated(vd.id, freshVd.id))
+        }
+        PiType(sparams, transform(to, senv)).copiedFrom(tpe)
+
+      case SigmaType(params, to) =>
+        val (sparams, senv) = params.foldLeft((Seq[t.ValDef](), env)) {
+          case ((sparams, env), vd) =>
+            val freshVd = vd.freshen
+            (sparams :+ transform(freshVd, env), env.updated(vd.id, freshVd.id))
+        }
+        SigmaType(sparams, transform(to, senv)).copiedFrom(tpe)
+
+      case _ => super.transform(tpe, env)
+    }
+  }
+
+  val freshenerNoChooses = new Freshener(false)
```

Comment:
Micro-nit: I don't think you gain much by making these stable, I would suggest using a method instead to make the overrides in Stainless trees a bit nicer:
```scala
protected def freshener(freshenChooses: Boolean): Freshener = new Freshener(freshenChooses)
```

Reply:
Fixed, thanks! Should I even keep this function? (I called `new Freshener` directly from `freshenLocals`)

Reply:
Since the implementation of `freshenLocals` is so simple, I guess there isn't really any point in having an intermediate function. Creating the `Freshener` directly and overriding `freshenLocals` in Stainless sounds good.

Reply:
Oh I thought overriding/extending the class would be enough but you're right I'll have to override `freshenLocals` (which is ok). It should be good now then, thanks!